src/telit_he910.cpp

```diff
@@ -0,0 +1,2102 @@
+#include <string.h>
+#include <stdio.h>
+
+#ifdef TELIT_HE910_SUPPORT
+
+namespace gpio = openxc::gpio;
+namespace uart = openxc::interface::uart;
+namespace can = openxc::can; // using this to get publishVehicleMessage() for GPS...no sense re-inventing the wheel
+namespace http = openxc::http;
+namespace telit = openxc::telitHE910;
+namespace commands = openxc::commands;
+
+using openxc::interface::uart::UartDevice;
+using openxc::gpio::GpioValue;
+using openxc::gpio::GPIO_DIRECTION_OUTPUT;
+using openxc::gpio::GPIO_DIRECTION_INPUT;
+using openxc::gpio::GPIO_VALUE_HIGH;
+using openxc::gpio::GPIO_VALUE_LOW;
+using openxc::util::time::delayMs;
+using openxc::util::log::debug;
+using openxc::util::time::uptimeMs;
+using openxc::config::getConfiguration;
+using openxc::telitHE910::TELIT_CONNECTION_STATE;
+using openxc::payload::PayloadFormat;
+
+/*PRIVATE MACROS*/
+
+#define TELIT_MAX_MESSAGE_SIZE		512
+#define NETWORK_CONNECT_TIMEOUT		150000
+#define PDP_MAX_ATTEMPTS			3
+
+/*PRIVATE VARIABLES*/
+
+static unsigned int bauds[3] = {
+	230400,
+	115200,
+	57600
+};
+
+static const char* gps_fix_enum[openxc::telitHE910::FIX_MAX_ENUM] = {"NO_FIX_0", "NO_FIX", "2D_FIX", "3D_FIX"};
+
+static char recv_data[1024];	// common buffer for receiving data from the modem
+static char* pRx = recv_data;
+static TelitDevice* telitDevice;
+static bool connect = false;
+static const unsigned int sendBufferSize = 4096;
+static uint8_t sendBuffer[sendBufferSize];
+static uint8_t* pSendBuffer = sendBuffer;
+static const unsigned int commandBufferSize = 512;
+static uint8_t commandBuffer[commandBufferSize];
+static uint8_t* pCommandBuffer = commandBuffer;
+
+static TELIT_CONNECTION_STATE state = telit::POWER_OFF;
+
+/*PRIVATE FUNCTIONS*/
+
+static bool autobaud(openxc::telitHE910::TelitDevice* device);
+static void telit_setIoDirection(void);
+static void setPowerState(bool enable);
+static bool sendCommand(TelitDevice* device, const char* command, const char* response, uint32_t timeoutMs);
+static bool sendCommand(TelitDevice* device, const char* command, const char* response, const char* error, uint32_t timeoutMs);
+static void sendData(TelitDevice* device, char* data, unsigned int len);
+static void clearRxBuffer(void);
+static bool getResponse(const char* startToken, const char* stopToken, char* response, unsigned int maxLen);
+static bool parseGPSACP(const char* GPSACP);
+
+TELIT_CONNECTION_STATE openxc::telitHE910::getDeviceState() {
+	return state;
+}
+
+TELIT_CONNECTION_STATE openxc::telitHE910::connectionManager(TelitDevice* device) {
+
+	static unsigned int subState = 0;
+	static unsigned int timer = 0;
+	static unsigned int timeout = 0;
+	static NetworkDescriptor current_network = {};
+	static bool pdp_connected = false;
+	static uint8_t pdp_counter = 0;
+
+	// device stats
+	static unsigned int SIMstatus = 0;
+	static NetworkConnectionStatus connectStatus = UNKNOWN;
+	static char ICCID[32];
+	static char IMEI[32];
+	
+	switch(state)
```

Comment:
Is it possible to pull up the handlers for each state into a function? This switch is pretty big and hard to follow.